### PR TITLE
adding powerplatform_rest_query and powerplatform_rest

### PR DESCRIPTION
**Pull request by mawasile**
This pull request introduces changes to the PowerPlatform provider for Terraform. The changes primarily focus on the addition of REST query data sources, updates to the REST resource, and improvements to error handling. 

Addition of REST query data sources:

* [`docs/data-sources/rest_query.md`](diffhunk://#diff-cf7d713dae39e9c1ba5d6ca1e1ccaac9e331c7796de97bc8b05c49f4530e33a1R1-R82): Added documentation for the REST query data source. This data source fetches API requests.
* [`examples/data-sources/powerplatform_rest_query/data-source.tf`](diffhunk://#diff-7fbe046bceb134a3eeb967c403a344d114d15e5bb9d9d7eb749ebd32c9625af3R1-R30): Added an example of how to use the REST query data source.
* [`examples/data-sources/powerplatform_rest_query/output.tf`](diffhunk://#diff-3f1e0cf629ab595a16d140595bdb74cc48aabbd51a3ef99695107fe87d2753c2R1-R4): Added an example output for the REST query data source.
* [`internal/powerplatform/datasource_rest_query_test.go`](diffhunk://#diff-d5156a7778c4cf0bd6efa29d036e080e5fbac294693e053b45f43e57db5676b8R1-R146): Added unit and acceptance tests for the REST query data source.

Updates to the REST resource:

* [`docs/resources/rest.md`](diffhunk://#diff-05ebf31b8a962010445030fefd4f260b65588b56591cb57bd933867e20a7b6eeR1-R172): Updated the documentation for the REST resource, which executes web API requests.
* [`examples/resources/powerplatform_rest/resource.tf`](diffhunk://#diff-44d32a645b90e81723b832dac7e55cced55bfa7081dcdea0bf7eb214d26b3036R1-R36): Added an example of how to use the REST resource.
* [`examples/resources/powerplatform_rest/output.tf`](diffhunk://#diff-35367e98e3202f7eab6e5971dbd3784991751945e52219861c5dda92f7fbbb0dR1-R5): Added an example output for the REST resource.

Improvements to error handling:

* [`internal/powerplatform/api/api_client.go`](diffhunk://#diff-03bd6d50cb06de9a7c49a1106113e5d0628b0c8b679d1816973a3e60c741c999R14-R19): Enhanced the `Execute` method to handle different types of body content and to return more detailed error information. Introduced a new `ExecuteForGivenScope` method to allow execution of requests with a specified scope. [[1]](diffhunk://#diff-03bd6d50cb06de9a7c49a1106113e5d0628b0c8b679d1816973a3e60c741c999R14-R19) [[2]](diffhunk://#diff-03bd6d50cb06de9a7c49a1106113e5d0628b0c8b679d1816973a3e60c741c999L56-R121)
* [`internal/powerplatform/helpers/error.go`](diffhunk://#diff-f605334f02d384f0f610733a3bc70a367e1b3319002643f9a738360c9668454fR15-R16): Improved error handling by introducing new error codes and enhancing the `WrapIntoProviderError` function. [[1]](diffhunk://#diff-f605334f02d384f0f610733a3bc70a367e1b3319002643f9a738360c9668454fR15-R16) [[2]](diffhunk://#diff-f605334f02d384f0f610733a3bc70a367e1b3319002643f9a738360c9668454fR56-R67)

Minor changes:

* `docs/resources/data_record.md`, `examples/resources/powerplatform_data_record/resource.tf`: Updated the data records example to use the environment ID from the `powerplatform_environment` resource instead of a hardcoded value. [[1]](diffhunk://#diff-0355f94f8ebd58f4747c74d9279cea1378d8260d340dc453a1f9635140b37007L29-L35) [[2]](diffhunk://#diff-0355f94f8ebd58f4747c74d9279cea1378d8260d340dc453a1f9635140b37007R40-R46) [[3]](diffhunk://#diff-63ab86988d1f28001ec761b6542b9ff252eaca42e6a4593b6a8f16f38e365158L14-L20) [[4]](diffhunk://#diff-63ab86988d1f28001ec761b6542b9ff252eaca42e6a4593b6a8f16f38e365158R25-R31)

### CODE REVIEW

File: docs/data-sources/rest_query.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  dataverse = {
+    language_code     = "1033"
+    currency_code     = "USD"
+    security_group_id = "00000000-0000-0000-0000-000000000000"
+  }
+}
+
+data "powerplatform_rest_query" "webapi_query" {
+  environment_id       = powerplatform_environment.development.id
+  url                  = "/api/data/v9.2/RetrieveCurrentOrganization(AccessType=@p1)?@p1=Microsoft.Dynamics.CRM.EndpointAccessType'Default'"
```
[the remark this answers is not in the document]
Comment (mawasile) [reply docs/data-sources/rest_query.md:42]: thanks, forgot to regenerate the docs. BTW `make userdocs` should be something that we should run in a pipeline and checkit in, or block PR is there is a code change

File: docs/resources/rest.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  environment_type = "Sandbox"
+  dataverse = {
+    language_code     = "1033"
+    currency_code     = "USD"
+    security_group_id = "00000000-0000-0000-0000-000000000000"
+  }
+}
+
+resource "powerplatform_rest" "install_sample_data" {
+  environment_id = powerplatform_environment.env.id
```
[the remark this answers is not in the document]
Comment (mawasile) [reply docs/resources/rest.md:40]: fixed, userdocs where not updated

**Review by eduardodfmex — APPROVED** [review]

File: docs/resources/rest.md [hunk 3]
```diff
@@ -0,0 +1,191 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "powerplatform_rest Resource - powerplatform"
+subcategory: ""
+description: |-
+  Resource to execute web api requests. There are four distinct operations, that you can define idepenetly. The HTTP response' body of the operation, that was called as last, will be returned in 'output.body' \n\n:
```
Comment (mattdot) [inline docs/resources/rest.md:6]:
```suggestion
  Resource to execute web api requests. There are four distinct operations, that you can define independently. The HTTP response' body of the operation, that was called as last, will be returned in 'output.body' \n\n:
```

File: docs/resources/rest.md [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  * Create: will be called once during the lifecycle of the resource (first 'terraform apply')
+  * Read: terraform will call this operation every time during 'plan' and 'apply' to get the current state of the resource
+  * Update: will be called every time during 'terraform apply' if the resource has changed (change done by the user or different values returned by the 'read' operation than those in the current state)
+  * Destroy: will be called once during the lifecycle of the resource (last 'terraform destroy')
+  \n\nYOu don't have to define all the operations but there are some things to consider:
```
Comment (mattdot) [inline docs/resources/rest.md:11]:
```suggestion
  \n\nYou don't have to define all the operations but there are some things to consider:
```

File: docs/resources/rest.md [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+  * lack of 'create' operation will result in no reasource being created. If you only need to read values consider using datasource 'powerplatform_rest_query' instead
+  * lack of 'read' operation will result in no resource changes being tracked. That means that the 'update' operation will never be called
+  * lack of destroy will couse, that the resource will not be deleted during 'terraform destroy'
```
Comment (mattdot) [inline docs/resources/rest.md:14]:
```suggestion
  * lack of destroy will cause that the resource to not be deleted during 'terraform destroy'
```

**Review by mattdot — DISMISSED** [review]

**Review by mattdot — APPROVED** [review]

**Review by eduardodfmex — APPROVED** [review]

**Pull request merged**